lib/auth/machineid/machineidv1/spiffe_federation_syncer.go

```diff
@@ -0,0 +1,582 @@
+func NewSPIFFEFederationSyncer(cfg SPIFFEFederationSyncerConfig) (*SPIFFEFederationSyncer, error) {
+	switch {
+	case cfg.Backend == nil:
+		return nil, trace.BadParameter("backend: must be non-nil")
+	case cfg.Store == nil:
+		return nil, trace.BadParameter("store: must be non-nil")
+	case cfg.Logger == nil:
+		return nil, trace.BadParameter("logger: must be non-nil")
+	case cfg.Clock == nil:
+		return nil, trace.BadParameter("clock: must be non-nil")
+	}
+	if cfg.MinSyncInterval == 0 {
+		cfg.MinSyncInterval = minRefreshInterval
+	}
+	if cfg.MaxSyncInterval == 0 {
+		cfg.MaxSyncInterval = maxRefreshInterval
+	}
+	if cfg.DefaultSyncInterval == 0 {
+		cfg.DefaultSyncInterval = defaultRefreshInterval
+	}
+	if cfg.SyncTimeout == 0 {
+		cfg.SyncTimeout = defaultSyncTimeout
+	}
+	return &SPIFFEFederationSyncer{
+		cfg: cfg,
+	}, nil
+}
+
+func (s *SPIFFEFederationSyncer) Run(ctx context.Context) error {
+	// TODO: Should this go into a loop?
+	return backend.RunWhileLocked(ctx, backend.RunWhileLockedConfig{
+		// TODO: Evaluate sensible TTL/retry
+		LockConfiguration: backend.LockConfiguration{
+			Backend:       s.cfg.Backend,
+			LockName:      "spiffe_federation_syncer",
+			TTL:           time.Minute,
+			RetryInterval: time.Minute,
+		},
+		RefreshLockInterval: time.Second * 15,
+	}, s.runWhileLocked)
+}
+
+type trustDomainSyncState struct {
+	// eventsCh is an unbuffered channel for passing events to a specific
+	// SPIFFEFederations syncer.
+	eventsCh chan types.Event
+}
+
+// runWhileLocked is the core loop of the syncer that runs on a single auth
+// server.
+//
+// Its goal is to sync the contents of trust bundles from remote clusters to
+// the local cluster. It does this by creating a goroutine that manages each
+// federated cluster.
+func (s *SPIFFEFederationSyncer) runWhileLocked(ctx context.Context) error {
+	s.cfg.Logger.InfoContext(
+		ctx,
+		"Obtained lock, SPIFFEFederation syncer is starting",
+	)
+	defer func() {
+		s.cfg.Logger.InfoContext(
+			ctx, "SPIFFEFederation syncer has stopped",
+		)
+	}()
+
+	// This wg will track all active syncers. We'll wait here until we're done.
+	wg := &sync.WaitGroup{}
+	defer func() {
+		wg.Wait()
+	}()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Set up state management that will hold a list of all active trust domain
+	// syncers.
+	states := map[string]trustDomainSyncState{}
+	mu := &sync.Mutex{}
+	startSyncingFederation := func(trustDomain string) {
+		mu.Lock()
+		defer mu.Unlock()
+
+		// Don't start if we're already syncing this trust domain.
+		_, ok := states[trustDomain]
+		if ok {
+			return
+		}
+
+		states[trustDomain] = trustDomainSyncState{
+			eventsCh: make(chan types.Event),
+		}
+
+		wg.Add(1)
+		go func() {
+			defer func() {
+				mu.Lock()
+				delete(states, trustDomain)
+				mu.Unlock()
+				wg.Done()
+			}()
+			s.syncTrustDomainLoop(ctx, trustDomain, states[trustDomain].eventsCh)
+		}()
```

Comment:
What would you say about leveraging `singleflight` to ensure that there is only one ongoing operation triggered for a particular key? Example: https://pkg.go.dev/golang.org/x/sync/singleflight#example-Group

Then the code can just process domains sequential and instead of relying on syncTrustDomainLoop -> nextSync you can just restart/recreate the Domain Watcher and in the result all domains will be periodically updated.